// @react-md/utils/lib/positioning/constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.BELOW_RIGHT_ANCHOR = exports.BELOW_INNER_RIGHT_ANCHOR = exports.BELOW_CENTER_ANCHOR = exports.BELOW_INNER_LEFT_ANCHOR = exports.BELOW_LEFT_ANCHOR = exports.BOTTOM_RIGHT_ANCHOR = exports.BOTTOM_INNER_RIGHT_ANCHOR = exports.BOTTOM_CENTER_ANCHOR = exports.BOTTOM_INNER_LEFT_ANCHOR = exports.BOTTOM_LEFT_ANCHOR = exports.CENTER_RIGHT_ANCHOR = exports.CENTER_INNER_RIGHT_ANCHOR = exports.CENTER_CENTER_ANCHOR = exports.CENTER_INNER_LEFT_ANCHOR = exports.CENTER_LEFT_ANCHOR = exports.TOP_RIGHT_ANCHOR = exports.TOP_INNER_RIGHT_ANCHOR = exports.TOP_CENTER_ANCHOR = exports.TOP_INNER_LEFT_ANCHOR = exports.TOP_LEFT_ANCHOR = exports.ABOVE_RIGHT_ANCHOR = exports.ABOVE_INNER_RIGHT_ANCHOR = exports.ABOVE_CENTER_ANCHOR = exports.ABOVE_INNER_LEFT_ANCHOR = exports.ABOVE_LEFT_ANCHOR = void 0;
exports.ABOVE_LEFT_ANCHOR = {
    x: "left",
    y: "above",
};
exports.ABOVE_INNER_LEFT_ANCHOR = {
    x: "inner-left",
    y: "above",
};
exports.ABOVE_CENTER_ANCHOR = {
    x: "center",
    y: "above",
};
exports.ABOVE_INNER_RIGHT_ANCHOR = {
    x: "inner-right",
    y: "above",
};
exports.ABOVE_RIGHT_ANCHOR = {
    x: "right",
    y: "above",
};
exports.TOP_LEFT_ANCHOR = {
    x: "left",
    y: "top",
};
exports.TOP_INNER_LEFT_ANCHOR = {
    x: "inner-left",
    y: "top",
};
exports.TOP_CENTER_ANCHOR = {
    x: "center",
    y: "top",
};
exports.TOP_INNER_RIGHT_ANCHOR = {
    x: "inner-right",
    y: "top",
};
exports.TOP_RIGHT_ANCHOR = {
    x: "right",
    y: "top",
};
exports.CENTER_LEFT_ANCHOR = {
    x: "left",
    y: "center",
};
exports.CENTER_INNER_LEFT_ANCHOR = {
    x: "inner-left",
    y: "center",
};
exports.CENTER_CENTER_ANCHOR = {
    x: "center",
    y: "center",
};
exports.CENTER_INNER_RIGHT_ANCHOR = {
    x: "inner-right",
    y: "center",
};
exports.CENTER_RIGHT_ANCHOR = {
    x: "right",
    y: "center",
};
exports.BOTTOM_LEFT_ANCHOR = {
    x: "left",
    y: "bottom",
};
exports.BOTTOM_INNER_LEFT_ANCHOR = {
    x: "inner-left",
    y: "bottom",
};
exports.BOTTOM_CENTER_ANCHOR = {
    x: "center",
    y: "bottom",
};
exports.BOTTOM_INNER_RIGHT_ANCHOR = {
    x: "inner-right",
    y: "bottom",
};
exports.BOTTOM_RIGHT_ANCHOR = {
    x: "right",
    y: "bottom",
};
exports.BELOW_LEFT_ANCHOR = {
    x: "left",
    y: "below",
};
exports.BELOW_INNER_LEFT_ANCHOR = {
    x: "inner-left",
    y: "below",
};
exports.BELOW_CENTER_ANCHOR = {
    x: "center",
    y: "below",
};
exports.BELOW_INNER_RIGHT_ANCHOR = {
    x: "inner-right",
    y: "below",
};
exports.BELOW_RIGHT_ANCHOR = {
    x: "right",
    y: "below",
};
