// @react-md/utils/lib/Dir.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Dir = exports.DEFAULT_DIR = exports.useDir = void 0;
var react_1 = __importStar(require("react"));
var context = react_1.createContext({
    root: true,
    dir: "ltr",
    toggleDir: function () {
        throw new Error("Tried to toggle the current writing direction without initializing the `Dir` component.");
    },
});
var Provider = context.Provider;
/**
 * Gets the writing direction context which provides access to the current `dir`
 * and a `toggleDir` function.
 *
 * @remarks \@since 2.3.0
 */
function useDir() {
    var _a = react_1.useContext(context), _root = _a.root, current = __rest(_a, ["root"]);
    return current;
}
exports.useDir = useDir;
/**
 * @remarks \@since 2.3.0
 */
var DEFAULT_DIR = function () {
    var dir = "ltr";
    if (typeof document !== "undefined") {
        var rootDir = document.documentElement.getAttribute("dir");
        dir = rootDir === "rtl" ? "rtl" : "ltr";
    }
    return dir;
};
exports.DEFAULT_DIR = DEFAULT_DIR;
/**
 * The `Dir` component is used to handle the current writing direction within
 * your app as well as conditionally updating the writing direction for small
 * sections in your app. When this component is used for the first time near the
 * root of your React component tree, the current direction will be applied to
 * the root `<html>` element. Otherwise the current dir will be cloned into the
 * child element so it can be passed as a prop.
 *
 * ```tsx
 * // html element will be updated to have `dir="ltr"`
 * ReactDOM.render(<Dir><App /></Dir>, root)
 * ```
 *
 * ```tsx
 * // html element will be updated to have `dir="rtl"` while the `<span>` will
 * // now be `<span dir="ltr">`
 * ReactDOM.render(
 *   <Dir defaultDir="rtl">
 *     <Some>
 *       <Other>
 *         <Components>
 *           <Dir defaultDir="ltr">
 *             <span>Content</span>
 *           </Dir>
 *         </Components>
 *       </Other>
 *     </Some>
 *   </Dir>,
 *   root
 * );
 * ```
 *
 * Note: Since the `dir` is cloned into the child element, you need to make sure
 * that the child is either a DOM element or the `dir` prop is passed from your
 * custom component.
 *
 * @remarks \@since 2.3.0
 */
function Dir(_a) {
    var children = _a.children, _b = _a.defaultDir, defaultDir = _b === void 0 ? exports.DEFAULT_DIR : _b;
    var root = react_1.useContext(context).root;
    var _c = react_1.useState(defaultDir), dir = _c[0], setDir = _c[1];
    react_1.useEffect(function () {
        if (!root || typeof document === "undefined") {
            return;
        }
        document.documentElement.setAttribute("dir", dir);
        return function () {
            document.documentElement.removeAttribute("dir");
        };
    }, [dir, root]);
    var toggleDir = react_1.useCallback(function () {
        setDir(function (prevDir) { return (prevDir === "ltr" ? "rtl" : "ltr"); });
    }, []);
    var value = react_1.useMemo(function () { return ({ root: false, dir: dir, toggleDir: toggleDir }); }, [dir, toggleDir]);
    var child = react_1.Children.only(children);
    if (!root) {
        child = react_1.cloneElement(child, { dir: dir });
    }
    return react_1.default.createElement(Provider, { value: value }, child);
}
exports.Dir = Dir;
/* istanbul ignore next */
if (process.env.NODE_ENV !== "production") {
    context.displayName = "WritingDirection";
    try {
        var PropTypes = require("prop-types");
        Dir.propTypes = {
            children: PropTypes.element.isRequired,
            defaultDir: PropTypes.oneOfType([
                PropTypes.func,
                PropTypes.oneOf(["ltr", "rtl"]),
            ]),
        };
    }
    catch (e) { }
}
