// @react-md/table/lib/TableCellContent.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.TableCellContent = void 0;
var react_1 = __importStar(require("react"));
var classnames_1 = __importDefault(require("classnames"));
var button_1 = require("@react-md/button");
var icon_1 = require("@react-md/icon");
var utils_1 = require("@react-md/utils");
var block = utils_1.bem("rmd-table-cell");
/**
 * This is mostly an internal component since it is automatically used within
 * the `TableCell` component but this will conditionally wrap the `children`
 * within an `UnstyledButton` to make a clickable cell. This is really to help
 * with sort behavior within headers.
 */
exports.TableCellContent = react_1.forwardRef(function TableCellContent(_a, ref) {
    var id = _a.id, propIcon = _a.icon, style = _a.style, className = _a.className, sortOrder = _a.sortOrder, children = _a.children, propRotated = _a.rotated, props = __rest(_a, ["id", "icon", "style", "className", "sortOrder", "children", "rotated"]);
    if (!sortOrder || propIcon === null) {
        return react_1.default.createElement(react_1.default.Fragment, null, children);
    }
    var icon = null;
    if (sortOrder !== "none") {
        var rotated = propRotated !== null && propRotated !== void 0 ? propRotated : sortOrder === "descending";
        icon = react_1.default.createElement(icon_1.IconRotator, { rotated: rotated }, propIcon);
    }
    return (react_1.default.createElement(button_1.UnstyledButton, { id: id, ref: ref, style: style, className: classnames_1.default(block("child"), className) },
        react_1.default.createElement(icon_1.TextIconSpacing, __assign({}, props, { icon: icon }), children)));
});
/* istanbul ignore next */
if (process.env.NODE_ENV !== "production") {
    try {
        var PropTypes = require("prop-types");
        exports.TableCellContent.propTypes = {
            id: PropTypes.string,
            style: PropTypes.object,
            className: PropTypes.string,
            icon: PropTypes.node,
            sortOrder: PropTypes.oneOf(["ascending", "descending", "none", "other"]),
            rotated: PropTypes.bool,
            children: PropTypes.node,
        };
    }
    catch (e) { }
}
