// @react-md/progress/lib/CircularProgress.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.CircularProgress = void 0;
var react_1 = __importStar(require("react"));
var classnames_1 = __importDefault(require("classnames"));
var utils_1 = require("@react-md/utils");
var block = utils_1.bem("rmd-circular-progress");
exports.CircularProgress = react_1.forwardRef(function CircularProgress(_a, ref) {
    var className = _a.className, propSvgStyle = _a.svgStyle, svgClassName = _a.svgClassName, propCircleStyle = _a.circleStyle, circleClassName = _a.circleClassName, value = _a.value, _b = _a.min, min = _b === void 0 ? 0 : _b, _c = _a.max, max = _c === void 0 ? 100 : _c, _d = _a.radius, radius = _d === void 0 ? 30 : _d, _e = _a.center, center = _e === void 0 ? 33 : _e, _f = _a.viewBox, viewBox = _f === void 0 ? "0 0 66 66" : _f, _g = _a.dashoffset, dashoffset = _g === void 0 ? 187 : _g, _h = _a.animate, animate = _h === void 0 ? true : _h, _j = _a.centered, centered = _j === void 0 ? true : _j, _k = _a.maxRotation, maxRotation = _k === void 0 ? 360 * 1.75 : _k, _l = _a.small, small = _l === void 0 ? false : _l, props = __rest(_a, ["className", "svgStyle", "svgClassName", "circleStyle", "circleClassName", "value", "min", "max", "radius", "center", "viewBox", "dashoffset", "animate", "centered", "maxRotation", "small"]);
    var progress;
    if (typeof value === "number") {
        progress = utils_1.getPercentage(min, max, value);
    }
    var svgStyle = react_1.useMemo(function () {
        if (typeof progress !== "number") {
            return propSvgStyle;
        }
        var transform = propSvgStyle && propSvgStyle.transform;
        if (maxRotation > 0) {
            var rotate = "rotate(" + maxRotation * progress + "deg)";
            transform = "" + rotate + (transform ? " " + transform : "");
        }
        return __assign(__assign({}, propSvgStyle), { WebkitTransform: transform, transform: transform });
    }, [progress, maxRotation, propSvgStyle]);
    var circleStyle = react_1.useMemo(function () {
        if (typeof progress !== "number") {
            return propCircleStyle;
        }
        return __assign(__assign({}, propCircleStyle), { strokeDashoffset: dashoffset - dashoffset * progress });
    }, [progress, propCircleStyle, dashoffset]);
    var determinate = typeof progress === "number";
    var indeterminate = !determinate;
    return (react_1.default.createElement("span", __assign({}, props, { ref: ref, role: "progressbar", "aria-valuemin": min, "aria-valuemax": max, "aria-valuenow": value, className: classnames_1.default(block({ centered: centered, small: small }), className) }),
        react_1.default.createElement("svg", { style: svgStyle, className: classnames_1.default(block("svg", {
                animate: animate && determinate,
                determinate: determinate,
                indeterminate: indeterminate,
            }), svgClassName), viewBox: viewBox },
            react_1.default.createElement("circle", { style: circleStyle, className: classnames_1.default(block("circle", {
                    animate: animate && determinate,
                    determinate: determinate,
                    indeterminate: indeterminate,
                }), circleClassName), r: radius, cx: center, cy: center }))));
});
/* istanbul ignore next */
if (process.env.NODE_ENV !== "production") {
    try {
        var PropTypes = require("prop-types");
        exports.CircularProgress.propTypes = {
            id: PropTypes.string.isRequired,
            className: PropTypes.string,
            min: PropTypes.number,
            max: PropTypes.number,
            value: PropTypes.number,
            animate: PropTypes.bool,
            svgStyle: PropTypes.object,
            svgClassName: PropTypes.string,
            circleStyle: PropTypes.object,
            circleClassName: PropTypes.string,
            radius: PropTypes.number,
            center: PropTypes.number,
            centered: PropTypes.bool,
            maxRotation: PropTypes.number,
            dashoffset: PropTypes.number,
            small: PropTypes.bool,
            viewBox: PropTypes.string,
        };
    }
    catch (e) { }
}
