// @react-md/tree/lib/useTreeItemSelection.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.useTreeItemSelection = void 0;
var react_1 = require("react");
/**
 * A hook that implements the base functionality for selecting different tree
 * items.
 *
 * @param defaultSelectedIds - The default list of tree item ids that should be
 * expanded by default
 * @param multiSelect - Boolean if the tree can have multiple items selected or
 * not.
 * @returns an object containing props that can be passed to the `Tree`
 * component to handle the selection state within the tree
 */
function useTreeItemSelection(defaultSelectedIds, multiSelect) {
    if (multiSelect === void 0) { multiSelect = false; }
    var _a = react_1.useState(defaultSelectedIds), selectedIds = _a[0], setSelectedIds = _a[1];
    var onItemSelect = react_1.useCallback(function (itemId) {
        setSelectedIds(function (selectedIds) {
            if (!multiSelect) {
                if (selectedIds[0] === itemId && selectedIds.length === 1) {
                    return selectedIds;
                }
                return [itemId];
            }
            var i = selectedIds.indexOf(itemId);
            var nextSelectedIds = selectedIds.slice();
            if (i === -1) {
                nextSelectedIds.push(itemId);
            }
            else {
                nextSelectedIds.splice(i, 1);
            }
            return nextSelectedIds;
        });
    }, [multiSelect]);
    var onMultiItemSelect = react_1.useCallback(function (itemIds) {
        setSelectedIds(itemIds);
    }, []);
    return {
        selectedIds: selectedIds,
        multiSelect: multiSelect,
        onItemSelect: onItemSelect,
        onMultiItemSelect: onMultiItemSelect,
    };
}
exports.useTreeItemSelection = useTreeItemSelection;
