// @react-md/form/lib/toggle/AsyncSwitch.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.AsyncSwitch = void 0;
var react_1 = __importStar(require("react"));
var classnames_1 = __importDefault(require("classnames"));
var progress_1 = require("@react-md/progress");
var Switch_1 = require("./Switch");
// this is used while the loading state is enabled to "disable" the switch
// toggle. If we disable the entire switch, keyboard focus is lost which is not
// desired.
var noop = function () {
    // do nothing
};
/**
 * This component will create an async switch that will show a loading indicator
 * and prevent the switch from being toggled while the loading state is true.
 */
exports.AsyncSwitch = react_1.forwardRef(function AsyncSwitch(_a, ref) {
    var id = _a.id, disabled = _a.disabled, className = _a.className, progressStyle = _a.progressStyle, progressClassName = _a.progressClassName, loading = _a.loading, onChange = _a.onChange, props = __rest(_a, ["id", "disabled", "className", "progressStyle", "progressClassName", "loading", "onChange"]);
    var progressId = id + "-loading";
    return (react_1.default.createElement(Switch_1.Switch, __assign({}, props, progress_1.getProgressA11y(progressId, loading), { id: id, ref: ref, disabled: disabled, className: classnames_1.default("rmd-switch--async", className), labelDisabled: disabled || false, onChange: loading ? noop : onChange }), loading && (react_1.default.createElement(progress_1.CircularProgress, { id: progressId, style: progressStyle, className: classnames_1.default("rmd-switch__progress", progressClassName), centered: false }))));
});
/* istanbul ignore next */
if (process.env.NODE_ENV === "production") {
    try {
        var PropTypes = require("prop-types");
        exports.AsyncSwitch.propTypes = {
            id: PropTypes.string.isRequired,
            className: PropTypes.string,
            onChange: PropTypes.func,
            disabled: PropTypes.bool,
            loading: PropTypes.bool.isRequired,
            progressStyle: PropTypes.object,
            progressClassName: PropTypes.string,
        };
    }
    catch (e) { }
}
