// @react-md/tooltip/lib/constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TOOLTIP_SPACING_VAR = exports.DEFAULT_TOOLTIP_POSITION = exports.DEFAULT_TOOLTIP_MARGIN = exports.DEFAULT_TOOLTIP_DENSE_SPACING = exports.DEFAULT_TOOLTIP_SPACING = exports.DEFAULT_TOOLTIP_THRESHOLD = exports.DEFAULT_TOOLTIP_DELAY = exports.DEFAULT_TOOLTIP_TIMEOUT = exports.DEFAULT_TOOLTIP_CLASSNAMES = void 0;
exports.DEFAULT_TOOLTIP_CLASSNAMES = {
    appear: "rmd-tooltip--enter",
    appearActive: "rmd-tooltip--visible",
    enter: "rmd-tooltip--enter",
    enterActive: "rmd-tooltip--visible",
    enterDone: "rmd-tooltip--visible",
    exit: "rmd-tooltip--visible rmd-tooltip--exit",
    exitActive: "rmd-tooltip--exit-active",
};
exports.DEFAULT_TOOLTIP_TIMEOUT = {
    enter: 200,
    exit: 150,
};
exports.DEFAULT_TOOLTIP_DELAY = 1000;
exports.DEFAULT_TOOLTIP_THRESHOLD = 0.75;
/** @remarks \@since 2.8.0 */
exports.DEFAULT_TOOLTIP_SPACING = "1.5rem";
/** @remarks \@since 2.8.0 */
exports.DEFAULT_TOOLTIP_DENSE_SPACING = "0.875rem";
/** @remarks \@since 2.8.0 */
exports.DEFAULT_TOOLTIP_MARGIN = 16;
/** @remarks \@since 2.8.0 */
exports.DEFAULT_TOOLTIP_POSITION = "below";
/**
 * @internal
 * @remarks \@since 2.8.0
 */
exports.TOOLTIP_SPACING_VAR = "--rmd-tooltip-spacing";
