// @react-md/layout/lib/LayoutAppBar.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.LayoutAppBar = void 0;
var react_1 = __importStar(require("react"));
var app_bar_1 = require("@react-md/app-bar");
var LayoutAppBarTitle_1 = require("./LayoutAppBarTitle");
var LayoutNavToggle_1 = require("./LayoutNavToggle");
var LayoutProvider_1 = require("./LayoutProvider");
/**
 * This is the default implementation for an `AppBar` within the `Layout` that
 * will conditionally render the default `LayoutNavToggle` button and
 * `AppBarTitle` depending on specific props that were provided.
 */
exports.LayoutAppBar = react_1.forwardRef(function LayoutAppBar(_a, ref) {
    var children = _a.children, _b = _a.fixed, fixed = _b === void 0 ? true : _b, propNavToggle = _a.navToggle, navToggleProps = _a.navToggleProps, customTitle = _a.customTitle, titleChildren = _a.title, titleProps = _a.titleProps, props = __rest(_a, ["children", "fixed", "navToggle", "navToggleProps", "customTitle", "title", "titleProps"]);
    var _c = LayoutProvider_1.useLayoutConfig(), baseId = _c.baseId, layout = _c.layout;
    var nav = propNavToggle;
    if (typeof nav === "undefined") {
        // set the key to the current layout since we want the button to re-mount
        // on layout changes so the transition does not occur
        nav = react_1.default.createElement(LayoutNavToggle_1.LayoutNavToggle, __assign({ key: layout }, navToggleProps));
    }
    var title = customTitle;
    if (typeof title === "undefined") {
        title = (react_1.default.createElement(LayoutAppBarTitle_1.LayoutAppBarTitle, __assign({}, titleProps), titleChildren));
    }
    return (react_1.default.createElement(app_bar_1.AppBar, __assign({ id: baseId + "-header" }, props, { ref: ref, fixed: fixed }),
        nav,
        title,
        children));
});
/* istanbul ignore next */
if (process.env.NODE_ENV !== "production") {
    try {
        var PropTypes = require("prop-types");
        exports.LayoutAppBar.propTypes = {
            fixed: PropTypes.bool,
            navToggle: PropTypes.node,
            navToggleProps: PropTypes.object,
            className: PropTypes.string,
            customTitle: PropTypes.node,
            title: PropTypes.node,
            titleProps: PropTypes.object,
            children: PropTypes.node,
        };
    }
    catch (error) { }
}
